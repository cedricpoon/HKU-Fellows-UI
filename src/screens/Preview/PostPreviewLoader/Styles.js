import { StyleSheet } from 'react-native';
import { gridBase } from 'hkufui/theme/grid';
import { seperator } from 'hkufui/theme/palette';

export default StyleSheet.create({
  placeholderGroup: {
    marginLeft: gridBase * 7,
    marginTop: gridBase,
    marginBottom: gridBase * 2,
    marginRight: gridBase * 4
  },
  placeholder: {
    height: gridBase * 2,
    marginTop: gridBase * 2,
    alignSelf: 'flex-start',
    justifyContent: 'center',
    backgroundColor: seperator,
    borderRadius: gridBase / 2
  },
  placeholderShort: {
    width: '30%'
  },
  placeholderLong: {
    width: '100%'
  },
  gradient: {
    flex: 1,
    width: gridBase * 25
  },
  hr: {
    height: 1,
    backgroundColor: seperator,
    marginTop: gridBase,
    width: '100%',
    marginLeft: gridBase * 7,
  }
});
