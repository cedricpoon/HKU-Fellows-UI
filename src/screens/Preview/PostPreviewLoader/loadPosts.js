import { FILL_POSTS, LOAD_POSTS, LOAD_POSTS_FAILED } from "hkufui/src/constants/actionTypes";
import * as status from 'hkufui/src/constants/loadStatus';

import postList from 'hkufui/mock/public/posts';

export function fetchPostsSafe(callforth) {
  return (dispatch, getState) => {
    const { posts } = getState();
    if (posts && posts.status !== status.LOADING) {
      dispatch(callforth);
      dispatch(fetchPosts());
    }
  };
}

export function fetchPosts() {
  const latency = Math.floor(Math.random() * 5 + 1) * 1000;
  const failed = Math.floor(Math.random() * 5 + 1) === 5;
  const empty = Math.floor(Math.random() * 5 + 1) === 5;

  return (dispatch) => {
    dispatch(onLoad());

    // mocking of fetch from WebAPI
    setTimeout(() => {
      if (!failed)
        dispatch(onFill(empty ? [] : postList));
      else
        dispatch(onFail());
    }, latency);
  };
}

export const onFill = (posts) => ({
  type: FILL_POSTS,
  payload: {
    posts: posts
  }
});

export const onLoad = () => ({
  type: LOAD_POSTS
});

export const onFail = () => ({
  type: LOAD_POSTS_FAILED
});

const handleLoadPosts = (state = {}, action = {}) => {
  switch (action.type) {
    case FILL_POSTS:
      return {
        ...state,
        status: status.OK,
        posts: action.payload.posts
      }
    case LOAD_POSTS:
      return {
        ...state,
        status: status.LOADING,
        posts: []
      }
    case LOAD_POSTS_FAILED:
      return {
        ...state,
        status: status.FAIL,
        posts: []
      }
    default:
      return state;
  }
};

export default handleLoadPosts;
