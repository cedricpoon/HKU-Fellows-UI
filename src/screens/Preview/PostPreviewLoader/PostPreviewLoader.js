import React, { Component } from 'react';
import { Dimensions } from 'react-native';
import { View } from 'native-base';
import {
  PlaceholderContainer,
  Placeholder
} from 'react-native-loading-placeholder';
import LinearGradient from 'react-native-linear-gradient';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  PostScrollable,
  PostPlaceholder
} from 'hkufui/components';
import { OK, FAIL, LOADING, STILL } from 'hkufui/src/constants/loadStatus';
import { fetchPosts } from './loadPosts';
import { localize } from 'hkufui/locale';
import styles from './Styles';

const locale = localize({ language: 'en', country: 'hk' });

const LOADER_GROUP_HEIGHT = 80;
const EST_HEADER_HEIGHT = 65;
const EST_FOOTER_HEIGHT = 55;
const LIGHTER_COLOR = '#eeeeee';
const DARKER_COLOR = '#dddddd';

export class PostPreviewLoader extends Component {

  componentDidMount() {
    const { location, status, onFetchPosts } = this.props;

    // load on reopen app
    if (location !== '' && status === STILL) {
      onFetchPosts();
    }
  }

  _renderGradient() {
    return (
      <LinearGradient
        colors={[LIGHTER_COLOR, DARKER_COLOR, LIGHTER_COLOR]}
        start={{ x: 1.0, y: 0.0 }}
        end={{ x: 0.0, y: 0.0 }}
        style={styles.gradient}
      />
    );
  }

  _renderLoadPlaceholder() {
    const { height } = Dimensions.get('window');
    let context = [];

    for (let i = 0; i < height - EST_FOOTER_HEIGHT - EST_HEADER_HEIGHT; i += LOADER_GROUP_HEIGHT) {
      context.push(
        <View key={i} style={styles.placeholderGroup}>
          <Placeholder
            style={[styles.placeholder, styles.placeholderShort]}
          />
          <Placeholder
            style={[styles.placeholder, styles.placeholderLong]}
          />
          <Placeholder
            style={[styles.placeholder, styles.placeholderLong]}
          />
        </View>
      );
      context.push(<View key={`${i}s`} style={styles.hr}></View>);
    }

    return(
      <PlaceholderContainer
        animatedComponent={this._renderGradient()}
        duration={1000}
        delay={1000}
      >
        {context}
      </PlaceholderContainer>
    );
  }

  render() {
    const { posts, location, status, onFetchPosts } = this.props;

    if (posts && posts.length > 0 && status === OK) {
      return (<PostScrollable posts={posts} />);
    } else if (location === '' && status === STILL) {
      return (
        <PostPlaceholder
          headline={locale['post.noCourseTitle']}
          headlineColor='green'
          subHeadline={locale['post.noCourseContent']}
          icon={{name: 'md-book'}}
        />
      );
    } else if (posts && posts.length === 0 && status === OK) {
      return (
        <PostPlaceholder
          headline={locale['post.noPostTitle']}
          subHeadline={locale['post.noPostContent']}
          icon={{name: 'paper'}}
          button={{text: locale['post.noPostButton']}}
        />
      );
    } else if (status === FAIL) {
      return (
        <PostPlaceholder
          headline={locale['post.loadErrorTitle']}
          headlineColor='red'
          subHeadline={locale['post.loadErrorContent']}
          icon={{name: 'flash', active: true}}
          button={{text: locale['post.loadErrorButton'], onPress: onFetchPosts}}
        />
      );
    } else {
      return (this._renderLoadPlaceholder());
    }
  }
}

PostPreviewLoader.defaultProps = {
  posts: [],
  location: '',
  status: STILL
}

PostPreviewLoader.propTypes = {
  posts: PropTypes.arrayOf(PropTypes.object).isRequired,
  location: PropTypes.string,
  status: PropTypes.oneOf([ OK, FAIL, LOADING, STILL ]),
  onFetchPosts: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  location: state.location.courseTitle,
  posts: state.posts.posts,
  status: state.posts.status
});

const mapDispatchToProps = dispatch => ({
  onFetchPosts: () => dispatch(fetchPosts())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PostPreviewLoader);
