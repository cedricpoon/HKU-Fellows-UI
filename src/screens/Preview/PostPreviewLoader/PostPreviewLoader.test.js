import React from 'react';
import { shallow } from 'enzyme';
import { PostPreviewLoader } from './PostPreviewLoader';

describe('Testing PostPreviewLoader', () => {
  const defaultProps = {
    courses: [],
    onFetchPosts: () => {}
  }

  it('renders as expected with defaultProps', () => {
    const wrapper = shallow(
      <PostPreviewLoader {...defaultProps} />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
