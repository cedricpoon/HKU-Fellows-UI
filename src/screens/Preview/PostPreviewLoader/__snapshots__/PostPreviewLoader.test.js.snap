// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing PostPreviewLoader renders as expected with defaultProps 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostPreviewLoader
    courses={Array []}
    location=""
    onFetchPosts={[Function]}
    posts={Array []}
    status="STILL"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "headline": "Welcome to HKU Fellows",
      "headlineColor": "green",
      "icon": Object {
        "name": "md-book",
      },
      "subHeadline": "Enjoy your peer learning journey with us",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "headline": "Welcome to HKU Fellows",
        "headlineColor": "green",
        "icon": Object {
          "name": "md-book",
        },
        "subHeadline": "Enjoy your peer learning journey with us",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing PostPreviewLoader renders as expected with failed state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostPreviewLoader
    courses={Array []}
    location="location"
    onFetchPosts={[Function]}
    posts={Array []}
    status="FAIL"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "button": Object {
        "onPress": [Function],
        "text": "Refresh",
      },
      "headline": "Load Failure",
      "headlineColor": "red",
      "icon": Object {
        "active": true,
        "name": "flash",
      },
      "subHeadline": "Oops! We hope a page refresh helps",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "button": Object {
          "onPress": [Function],
          "text": "Refresh",
        },
        "headline": "Load Failure",
        "headlineColor": "red",
        "icon": Object {
          "active": true,
          "name": "flash",
        },
        "subHeadline": "Oops! We hope a page refresh helps",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing PostPreviewLoader renders as expected with loading 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostPreviewLoader
    courses={Array []}
    location="location"
    onFetchPosts={[Function]}
    posts={Array []}
    status="LOADING"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "multiple": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "multiple": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing PostPreviewLoader renders as expected with no post placeholder 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostPreviewLoader
    courses={Array []}
    location="location"
    onFetchPosts={[Function]}
    posts={Array []}
    status="OK"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "button": Object {
        "text": "Compose",
      },
      "headline": "No Posts Yet",
      "icon": Object {
        "name": "md-paper",
      },
      "subHeadline": "Feel free to raise the first post",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "button": Object {
          "text": "Compose",
        },
        "headline": "No Posts Yet",
        "icon": Object {
          "name": "md-paper",
        },
        "subHeadline": "Feel free to raise the first post",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
